hq_superset/views.py

```diff
@@ -0,0 +1,162 @@
+        oauth_token = get_valid_cchq_oauth_token()
+        response = provider.get(datasource_list_url, token=oauth_token)
+        return self.render_template(
+            "hq_datasource_list.html",
+            datasources=response.json(),
+        )
+
+
+
+class CCHQApiException(Exception):
+    pass
+
+
+def refresh_hq_datasource(domain, datasource_id):
+    # This method pulls the data from CommCareHQ and creates/replaces the
+    #   corresponding Superset dataset
+    datasource_url = get_datasource_export_url(domain, datasource_id)
+    provider = superset.appbuilder.sm.oauth_remotes["commcare"]
+    oauth_token = get_valid_cchq_oauth_token()
+    response = provider.get(datasource_url, token=oauth_token)
+    if response.status_code != 200:
+        # Todo; logging
+        raise CCHQApiException("Error downloading the UCR export from HQ")
+    zipfile = ZipFile(BytesIO(response.content))
+    filename = zipfile.namelist()[0]
+    # Upload to table
+    database = get_ucr_database()
+    schema = get_schema_name_for_domain(domain)
+    csv_table = Table(table=datasource_id, schema=schema)
+
+    try:
+        df = pd.concat(
+            pd.read_csv(
+                chunksize=1000,
+                filepath_or_buffer=zipfile.open(filename),
+                encoding="utf-8",
+                # Todo; make date parsing work
+                parse_dates=True,
+                infer_datetime_format=True,
+                keep_default_na=True,
+            )
+        )
+
+        database.db_engine_spec.df_to_sql(
+            database,
+            csv_table,
+            df,
+            to_sql_kwargs={
+                "chunksize": 1000,
+                "if_exists": "replace",
+            },
+        )
+
+        # Connect table to the database that should be used for exploration.
+        # E.g. if hive was used to upload a csv, presto will be a better option
+        # to explore the table.
+        expore_database = database
+        explore_database_id = database.explore_database_id
+        if explore_database_id:
+            expore_database = (
+                db.session.query(Database)
+                .filter_by(id=explore_database_id)
+                .one_or_none()
+                or database
+            )
+
+        sqla_table = (
+            db.session.query(SqlaTable)
+            .filter_by(
+                table_name=csv_table.table,
+                schema=csv_table.schema,
+                database_id=expore_database.id,
+            )
+            .one_or_none()
+        )
+
+        if sqla_table:
+            sqla_table.fetch_metadata()
+        if not sqla_table:
+            sqla_table = SqlaTable(table_name=csv_table.table)
+            sqla_table.database = expore_database
+            sqla_table.database_id = database.id
+            sqla_table.user_id = g.user.get_id()
+            sqla_table.schema = csv_table.schema
+            sqla_table.fetch_metadata()
```

Comment:
It can be extracted out of this condition and then we can get rid of the first condition.

Reply:
https://github.com/dimagi/hq_superset/pull/7/files#r870067051